### ISSUE TITLE
Design of “within / bounded comparison operator” (a < b < c)

### ISSUE THREAD

**Issue by lucaswerkmeister** [ceylon/ceylon.ast#38]
I’m wondering how the “within” or “bounded comparison” operator (for `a < b < c`) should be designed.

In the RedHat compiler AST, you have a `class WithinOp(Term, Bound lowerBound, Bound upperBound)` with `class Bound(Term) of OpenBound|ClosedBound`. For example, the expression `a <= b < c` is a `WithinOp { b; lowerBound = ClosedBound(a); upperBound = OpenBound(c); }`.

There’s several things I don’t like about this:
- It feels weird that the same class is used for the `a <=` and `<= a` fragments.
- In `ceylon.ast`, because `Bound`s are not `ValueExpression`s (I think they’re `ExpressionIsh`s) I’d either have to widen the type of `Operation.children` (I already had to do that for `UnaryTypeOperation`s, and I hated it), or make `WithinOperation` (or whatever it’ll be called) not a subclass of `Operation` (direct subclass of `ValueExpression`).

I think I would prefer a design like

``` ceylon
class WithinOperation(
        lowerBound,
        subject /* or middle or whatever */,
        upperBound,
        Boolean lowerBoundOpen,
        Boolean upperBoundOpen)
```

(where `lowerBound`, `subject`, `upperBound` are all `Precedence10Expression`s), that is, have everything in one node, with two `Boolean`s to distinguish between `<` and `<=`. (All the names are made-up; there are probably better ones.)

However, that design doesn’t require any Ceylon features, and seems more or less obvious to me, so I wonder if there are any good reasons why it wasn’t used in the current compiler. Does anyone remember? (I believe @gavinking added this, or at least the typechecker part?)

(Note: I don’t think this issue is urgent; as far as I can tell, it shouldn’t be a problem if I continue adding higher-precedence operations and other nodes, and later add `WithinOperation` to the `Precedence11Expression` alias.)

PS: Another crazy idea would be to have a helper function that allowed you to write:

``` ceylon
// a < b <= c
value op = boundedComparison {
    a;
    isSmallerThan = b;
    whichIsSmallerThanOrEquals = c;
};
```

Basically, it has the parameters `-SmallerThan` (`<=`) and `-SmallerThanOrEquals` (`<`) each for `is-` (lower bound) and `whichIs-` (upper bound). Of course, this couldn’t be the actual class, because I can’t enforce its correct usage in the type system… but it might be a neat convenience function. Quick survey: Do you find the above code readable?
